dali/kernels/signal/resampling_gpu.cuh

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+#define DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/signal/resampling.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace resampling {
+
+struct SampleDesc {
+  void *out;
+  const void *in;
+  ResamplingWindow window;
+  int64_t in_len;  // num samples in input
+  int64_t out_len;  // num samples in output
+  int64_t nchannels;  // number of channels
+  double scale;  // in_sampling_rate / out_sampling_rate
+};
+
+/**
+ * @brief Resamples 1D signal (single or multi-channel), optionally converting to a different data type.
+ *
+ * @param samples sample descriptors
+ */
+template <typename Out, typename In, bool SingleChannel = false>
+__global__ void ResampleGPUKernel(const SampleDesc *samples) {
+  auto sample = samples[blockIdx.y];
+  double scale = sample.scale;
+  float fscale = scale;
+  int nchannels = SingleChannel ? 1 : sample.nchannels;
+  auto &window = sample.window;
+
+  Out* out = reinterpret_cast<Out*>(sample.out);
+  const In* in = reinterpret_cast<const In*>(sample.in);
+
+  int64_t grid_stride = static_cast<int64_t>(gridDim.x) * blockDim.x;
+  int64_t out_block = static_cast<int64_t>(blockIdx.x) * blockDim.x;
+  int64_t start_out_pos = out_block + threadIdx.x;
+
+  double in_block_f = out_block * scale;
+  int64_t in_block_i = std::floor(in_block_f);
+  float in_pos_start = in_block_f - in_block_i;
+  const In* in_blk_ptr = in + in_block_i * nchannels;
+
+  for (int64_t out_pos = start_out_pos; out_pos < sample.out_len;
+       out_pos += grid_stride, in_pos_start += fscale * grid_stride) {
```

Comment:
You're losing precision here. The CPU code advanced the position by one (scaled) sample. The position was calculated relative to the start of a block, thereby avoiding loss of precision. The number of times this position was advanced was limited to block_size and the maximum offset was block_size * scale. Here, you advance an unknown number of times and the range can span entire recording. This defeats the whole purpose of using double precision in the first place. The options are:
change the traversal order, so that a thread block processes adjacent blocks of data:
```
Block index for each output element - block size is 4, grid size is 3
[0 0 0 0 1 1 1 1 2 2 2 2 0 0 0 0 1 1 1 1 2 2 2 2 0 0 0 0]

Proposed order:
[ 0 0 0 0  0 0 0 0  1 1 1 1  1 1 1 1  1 1 1 1  2 2 2 2  2 2 2 2]
```
Also, we could try to measure the penalty of having the double precision calculations in each loop. Perhaps we can calculate the double precision numbers only in thread 0 of a warp and then use warp shuffling - this should help on machines with reduced fp64 performance (GTX/RTX/Titan).

Reply:
This a serious problem, not a "nice to have" - this makes it impossible to address individual source samples in recordings longer than 9 minutes at 16 kHz. Between 5 and 9 minutes it would decay to nearest neighbor resampling (yikes!) and 9 to 17 minutes we'd only be able to address even samples, after 17 minutes every fourth and so on. You get the picture.

Reply:
The problem should be resolved with the latest version